src/Bicep.Core/TypeSystem/TypeValidator.cs

```diff
@@ -148,12 +148,13 @@ public static TypeSymbol NarrowTypeAndCollectDiagnostics(ITypeManager typeManage
             }
 
             // TODO: The type of this expression and all subexpressions should be cached
-            TypeSymbol? expressionType = typeManager.GetTypeInfo(expression);
+            var expressionType = typeManager.GetTypeInfo(expression);
 
             // since we dynamically checked type, we need to collect the errors but only if the caller wants them
-            if (skipTypeErrors == false && expressionType is UnassignableTypeSymbol)
+            if (skipTypeErrors == false && expressionType is ErrorType)
             {
                 diagnostics.AddRange(expressionType.GetDiagnostics());
+                return targetType;
             }

```

Comment:
@majastrz I'm not quite sure why, but refactoring to remove the null return value forced me to return here. It feels like the right thing to do though, since continuing to type check on an error type feels odd. Do you know why we don't currently return here?

Reply:
In general, I guess it depends if it'll ever cause us to lose a meaningful error.

If `expressionType` is `ErrorType` then L165 will return it and add an error about type mismatch and the error message will mention the "error" type, so that's probably not super useful. I think returning early in L157 is the right thing to do also.

---
In reply to: [515416865](https://github.com/Azure/bicep/pull/753#discussion_r515416865) [](ancestors = 515416865)